File: 24. Ch-5-7.java
import java.util.*;

class TrieNode 
{
    Map<Character,TrieNode> children;
    TrieNode() 
    {
        children= new HashMap<>();
    }
}

class DistinctSubstrings 
{
    TrieNode root;
    DistinctSubstrings() 
    {
        root = new TrieNode();
    }

    public int countDistinct(String s) 
    {
        int count = 0;
        int n = s.length();
        for (int i = 0; i < n; i++) 
        {
            TrieNode node = root;
            for (int j = i; j < n; j++) 
            {
                char ch = s.charAt(j);
                if (!node.children.containsKey(ch)) 
                {
                    node.children.put(ch, new TrieNode());
                    count++;
                }
                node = node.children.get(ch);
            }
        }
        return count;
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        String s = sc.next();
        DistinctSubstrings ob = new DistinctSubstrings();
        System.out.println(ob.countDistinct(s));
    }
}

File: 24. Ch-5-7.py
class TrieNode:
    def __init__(self):
        self.children = {}

class DistinctSubstrings:
    def __init__(self):
        self.root = TrieNode()
    def count_subStrings(self, text):
        count = 0
        n = len(text)
        for i in range(n):
            node = self.root
            for j in range(i, n):
                ch = text[j]
                if ch not in node.children:
                    node.children[ch] = TrieNode()
                    count += 1
                node = node.children[ch]
        return count

s = input()
ob = DistinctSubstrings()
print(ob.count_subStrings(s))
